pallets/amm/src/lib.rs

```diff
@@ -0,0 +1,123 @@
+// Copyright 2021 Parallel Finance Developer.
+// This file is part of Parallel Finance.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! # Automatic Market Maker (AMM)
+//!
+//! Given any [X, Y] asset pair, "base" is the `X` asset while "quote" is the `Y` asset.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+extern crate alloc;
+
+mod pool_structs;
+
+pub use pallet::*;
+use pool_structs::{AMMCurve, LiquidityProviderAmounts, Pool, StabilityPool};
+
+#[frame_support::pallet]
+mod pallet {
+    use crate::{AMMCurve, LiquidityProviderAmounts, Pool, StabilityPool};
+    use core::marker::PhantomData;
+    use frame_support::{
+        pallet_prelude::{StorageDoubleMap, StorageMap, StorageValue, ValueQuery},
+        traits::{EnsureOrigin, GenesisBuild, Get, Hooks, IsType},
+        Blake2_128Concat, PalletId, Parameter,
+    };
+    use orml_traits::MultiCurrencyExtended;
+    use parallel_primitives::{Amount, Balance, CurrencyId, Rate};
```

Comment:
This approach is used across all Parallel pallets so it is necessary to talk with the team about such possibility

Reply:
Let's talk about it, I think Yubo is setting up a meeting. There aren't big implications code or functionality wise anyways for now but this may become troublesome in the future

Reply:
The primitive here works like a common type and helper functions registration across runtime and client. Similar to `sp_core` crate in Substrate. [Here](https://github.com/paritytech/substrate/blob/master/frame/collective/src/lib.rs#L45) is an example.